mir_perception/mir_object_segmentation/common/src/scene_segmentation.cpp

```diff
@@ -0,0 +1,181 @@
+
+    normal_estimation_omp_.setNumberOfThreads(4);
+    normal_estimation_omp_.setSearchMethod(boost::make_shared<pcl::search::KdTree<PointT> >());
+
+};
+SceneSegmentation::~SceneSegmentation()
+{
+
+};
+
+PointCloud::Ptr SceneSegmentation::segmentScene(const PointCloud::ConstPtr &cloud, std::vector<PointCloud::Ptr> &clusters,
+                                  std::vector<BoundingBox> &boxes, pcl::ModelCoefficients::Ptr &coefficients, double &workspace_height)
+{
+    PointCloud::Ptr filtered(new PointCloud);
+    PointCloud::Ptr plane(new PointCloud);
+    PointCloud::Ptr hull(new PointCloud);
+    pcl::PointIndices::Ptr segmented_cloud_inliers(new pcl::PointIndices);
+    std::vector<pcl::PointIndices> clusters_indices;
+
+    filtered = findPlane(cloud, hull, coefficients, workspace_height);
+
+    if (coefficients->values.size() == 0)
+    {
+        return  filtered;
+    }
+
+    extract_polygonal_prism_.setInputPlanarHull(hull);
+    extract_polygonal_prism_.setInputCloud(cloud);
+    extract_polygonal_prism_.setViewPoint(0.0, 0.0, 2.0);
+    extract_polygonal_prism_.segment(*segmented_cloud_inliers);
+
+    cluster_extraction_.setInputCloud(cloud);
+    cluster_extraction_.setIndices(segmented_cloud_inliers);
+    cluster_extraction_.extract(clusters_indices);
+
+    const Eigen::Vector3f normal(coefficients->values[0], coefficients->values[1], coefficients->values[2]);
+
+    for (size_t i = 0; i < clusters_indices.size(); i++)
+    {
+        const pcl::PointIndices& cluster_indices = clusters_indices[i];
+        PointCloud::Ptr cluster(new PointCloud);
+        pcl::copyPointCloud(*cloud, cluster_indices, *cluster);
+        clusters.push_back(cluster);
+        BoundingBox box = BoundingBox::create(cluster->points, normal);
+        boxes.push_back(box);
+    }
+    return filtered;
+}
+
+PointCloud::Ptr SceneSegmentation::findPlane(const PointCloud::ConstPtr &cloud, PointCloud::Ptr &hull,
+                              pcl::ModelCoefficients::Ptr &coefficients, double &workspace_height)
+{
+    PointCloud::Ptr filtered(new PointCloud);
+    PointCloud::Ptr plane(new PointCloud);
+    pcl::PointIndices::Ptr segmented_cloud_inliers(new pcl::PointIndices);
+
+    PointCloudN::Ptr normals(new PointCloudN);
+
+    voxel_grid_.setInputCloud(cloud);
+    voxel_grid_.filter(*filtered);
+
+    if (enable_passthrough_filter_)
+    {
+        pass_through_.setInputCloud(filtered);
+        pass_through_.filter(*filtered);
+    }
+
+    //normal_estimation_.setInputCloud(filtered);
+    //normal_estimation_.compute(*normals);
+
+    normal_estimation_omp_.setInputCloud(filtered);
+    normal_estimation_omp_.compute(*normals);
+
+    pcl::PointIndices::Ptr inliers(new pcl::PointIndices);
+
+    sac_.setModelType(pcl::SACMODEL_NORMAL_PARALLEL_PLANE);
+    sac_.setMethodType(pcl::SAC_RANSAC);
+    sac_.setAxis(Eigen::Vector3f(0.0, 0.0, 1.0));
+    //sac_.setAxis(Eigen::Vector3f(-0.866, 0.0, 0.5));
```

Comment:
was this for slanted planes? later we can make this a parameter I guess

Reply:
I am not sure about this. I dont remember why I came up with these numbers. I need to check it, or remove it.